modules/ml/src/main/java/org/apache/ignite/ml/inference/storage/model/LocalModelStorageProvider.java

```diff
@@ -48,8 +54,50 @@
     }
 
     /** {@inheritDoc} */
-    @Override public Lock lock(String key) {
-        Lock lock = new ReentrantLock();
-        return locks.computeIfAbsent(key, k -> new WeakReference<>(lock)).get();
+    @Override public synchronized Lock lock(String key) {
+        WeakReferenceWithCleanUp ref = locks.get(key);
+        try {
+            if (ref != null) {
+                Lock lockInRef = ref.get();
+
+                // Reference is not empty and it couldn't be emptied because object is reachable from "lockInRef".
+                if (lockInRef != null)
+                    return lockInRef;
+            }
+
+            // If reference doesn't exists or it's empty we create a new one.
+            Lock lock = new ReentrantLock();
+            locks.put(key, new WeakReferenceWithCleanUp(key, lock));
+            return lock;
+        }
+        finally {
+            // At this point we already replaced all keys we wanted to replace, so all empty references could be safely
+            // deleted.
+            while ((ref = (WeakReferenceWithCleanUp)refQueue.poll()) != null) {
+                String keyToBeRmv = ref.key;
+                // We double check that we don't replaced the key value already.
+                if (locks.containsKey(keyToBeRmv) && locks.get(keyToBeRmv).get() == null)
```

Comment:
can be replaced with locks.remove(ref.key, ref) ?

Reply:
Good point. Done.